doc/lsst.daf.butler/formatters.rst

```diff
@@ -0,0 +1,316 @@
+  This is because a user storing a dataset does not know which formatter the datastore has been configured to use or even if a dataset will be persisted.
+  For example, the caller has no real idea whether a particular compression option is supported or not because they have no visibility into whether the file written is in HDF5 or FITS or plain text.
+  For this reason write parameters are part of formatter configuration.
+
+Read Parameters and Read-Only Components
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Read parameters are usually applied during the retrieval of the associated Python type from the persisted artifact.
+This requires that the parameters are understood by that Python type.
+When read-only components involve read parameters there are now multiple ways in which the parameters can be applied.
+
+Consider the case of a pixel counting read-only component for an image dataset.
+A read parameter for subsetting the dataset should be applied to the image before the pixel counting is performed.
+It does not make sense for subsetting to be applied to the integer pixel count.
+
+For this reason read parameters for read-only components are processed prior to calculating the read-only component.
+
+Defining a Storage Class
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Storage classes are defined in the ``storageClasses`` section of the Butler configuration YAML file.
+A very simple declaration of a storage class with an associated python type looks like:
+
+.. code-block:: yaml
+
+   NumPixels:
+     pytype: int
+
+This declares that the ``NumPixels`` storage class is defined as a Python `int`.
+Nothing more is required for simple types.
+
+A composite storage class refers to a Python type that can be disassembled into distinct components that can be retrieved independently:
+
+.. code-block:: yaml
+
+  MaskedImage:
+    pytype: lsst.afw.image.MaskedImage
+    assembler: lsst.something.MaskedImageAssembler
+    parameters:
+      - subset
+    components:
+      image: Image
+      mask: Mask
+    readComponents:
+      npixels: NumPixels
+
+In this simplified definition for a masked image, there are two components declared along with a read-only component that returns the number of pixels in the image.
+The assembler should be able to disassemble the associated Python type into the ``image`` and ``mask`` components if the datastore requests disassembly.
+The assembler would also be used to process the ``subset`` read parameter if the formatter used by the datastore has declared it does not support the parameter.
+
+In some cases you may want to define specific storage classes that are specializations of a more generic definition.
+You can do this using YAML anchors and references but the preferred approach is to use the ``inheritsFrom`` key in the storage class definition:
+
+.. code-block:: yaml
+
+   MaskedImageI:
+     inheritsFrom: MaskedImage
+     pytype: lsst.afw.image.MaskedImageI
+     components:
+       image: ImageI
+       mask: MaskX
```

Comment:
Is there a reason this example doesn't match what's in the default config (which doesn't have the components)?

Reply:
Because I only wanted to show the essence of an example and not show a fully worked real life example (It probably doesn't have components because we probably don't use them anywhere).